packages/contracts/contracts/core/ManagedService.sol

```diff
@@ -0,0 +1,117 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import { AccessControl } from "@openzeppelin/contracts/access/AccessControl.sol";
+import { console } from "sphinx-forge-std/console.sol";
+
+/**
+ * @title ManagedService
+ * @notice Contract controlled by the Sphinx managed service. This contract is used by
+           the managed service to remotely execute deployments.
+           Users can opt in to this functionality if they choose to do so.
+ */
+contract ManagedService is AccessControl {
+    /**
+     * @dev Role required to make calls through this contract.
+     */
+    bytes32 public constant RELAYER_ROLE = keccak256("RELAYER_ROLE");
+
+    /**
+     * @notice Emitted when funds are withdrawn from this contract.
+     *
+     * @param recipient The recipient that withdrew the funds.
+     * @param amount    Amount of ETH withdrawn.
+     */
+    event Withdew(address indexed recipient, uint256 amount);
+
+    /**
+     * @notice Emitted when a call is made to a remote contract.
+     *
+     * @param relayer The address of the account that made the call.
+     * @param to      The address of the remote contract.
+     * @param data    A keccak256 hash of the input data.
+     */
+    event Called(address indexed relayer, address indexed to, bytes32 data);
+
+    /**
+     * @notice A modifer that refunds the caller for the gas spent in the function call.
+     */
+    modifier refund {
```

Comment:
nit: is it necessary to use a modifier, since we only attach it to a single function? Merging it into the `exec` function would be slightly cleaner imo because the `exec` function returns data (via `return res;`), but this doesn't actually exit the modifier. This is just a Solidity quirk, but it makes the flow of the function a little opaque

Reply:
This is intentional and makes it easier to accurately calculate the gas cost b/c the cost of the return value is included in the estimate calculated in the modifier.  Additionally, the use of a modifier is common practice for this use case, so I expect auditors to be familiar with the pattern or at least be able to find resources online related to it.